File: 0547-number-of-provinces/0547-number-of-provinces.cpp
class Solution {
    vector<int> vis;
    vector<vector<int>> adj;
    void dfs(int node)
    {
        vis[node] = 1;
        for(auto it : adj[node])
        {
            if(!vis[it])
            {
                dfs(it);
            }
        }
    }
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        vis.clear();
        adj.clear();
        int n = isConnected.size();
        vis.resize(n,0);
        adj.resize(n);
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<isConnected[i].size();j++)
            {
                if(isConnected[i][j])
                {
                    adj[i].push_back(j);
                    adj[j].push_back(i);
                }
            }
        }
        int province = 0;
        for(int i=0;i<n;i++)
        {
            if(!vis[i])
            {
                dfs(i);
                province++;
            }
        }
        return province;
    }
};

File: 0547-number-of-provinces/0547-number-of-provinces.py
class Solution:
    def __init__(self):
        self.adj = []
        self.vis = [0] * 201
        
    def dfs(self, node: int) -> None:
        self.vis[node] = 1
        for i in self.adj[node]:
            if self.vis[i] == 0:
                self.dfs(i)
        
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        self.adj = [[] for _ in range(n)]  # Initialize adjacency list for each node
        self.vis = [0] * n  # Adjust visited array to correct size
        
        # Build the adjacency list from the isConnected matrix
        for i in range(n):
            for j in range(n):
                if isConnected[i][j] == 1 and i != j:
                    self.adj[i].append(j)
                    self.adj[j].append(i)
                    
        count = 0
        # Perform DFS for each unvisited node
        for i in range(n):
            if self.vis[i] == 0:  # If node `i` is unvisited
                self.dfs(i)
                count += 1  # Increment province count
        
        return count

        
        